aten/src/TH/THMath.h

```diff
@@ -126,4 +126,70 @@ Date:  February 1996
 }
 #undef CENTRAL_RANGE
 
+static inline double TH_digamma(double x) {
+  double result = 0;
+  if (x < 0.5) {
+    result -= M_PI / tan(M_PI * x);
+    x = 1 - x;
+  }
+  for (int i = 0; i < 4; ++i) {
+    result -= 1 / x;
+    x += 1;
+  }
+  const double ixx = 1 / (x*x);
+  result += log(x) - 1 / (2*x) - ixx * (1./12 - ixx * (1./120 - ixx * (1./252)));
+  return result;
+}
+
+static inline float TH_digammaf(float x) {
+  float result = 0;
+  if (x < 0.5) {
+    result -= M_PI / tanf(M_PI * x);
+    x = 1 - x;
+  }
+  for (int i = 0; i < 4; ++i) {
+    result -= 1 / x;
+    x += 1;
+  }
+  const float ixx = 1 / (x*x);
+  result += logf(x) - 1 / (2*x) - ixx * (1.f/12 - ixx * (1.f/120 - ixx * (1.f/252)));
+  return result;
+}
+
+static inline double TH_trigamma(double x) {
+  double sign = +1;
+  double result = 0;
+  if (x < 0.5) {
+    sign = -1;
+    const double sin_pi_x = sin(M_PI * x);
+    result -= (M_PI * M_PI) / (sin_pi_x * sin_pi_x);
+    x = 1 - x;
+  }
+  for (int i = 0; i < 6; ++i) {
+    result += 1 / (x * x);
+    x += 1;
+  }
+  const double ixx = 1 / (x*x);
+  result += (1 + 1 / (2*x) + ixx * (1./6 - ixx * (1./30 - ixx * (1./42)))) / x;
+  return sign * result;
+}
+
+static inline float TH_trigammaf(float x) {
+  float sign = +1;
+  float result = 0;
+  if (x < 0.5) {
+    sign = -1;
+    const float sin_pi_x = sinf(M_PI * x);
+    result -= (M_PI * M_PI) / (sin_pi_x * sin_pi_x);
+    x = 1 - x;
+  }
+  for (int i = 0; i < 6; ++i) {
+    result += 1 / (x * x);
+    x += 1;
+  }
+  const float ixx = 1 / (x*x);
+  result += (1 + 1 / (2*x) + ixx * (1./6 - ixx * (1./30 - ixx * (1./42)))) / x;
```

Comment:
nit: `1./6` will force conversion to double, and I don't know if that's desired. I think it's fine to have only the double overloads unless this function gets auto-vectorized by the compiler (you can check that in objdump). Scalar double math is the same speed as float math on CPUs, and looking at the body of this function I doubt that the casts will become a bottleneck. For vector instructions you get 2x throughput, so it matters.

Reply:
Sorry my cuda skills are weak. How can I specify floating point constant of type `real`? I wrote one version that cast constants everywhere like `(real)(1./6)` but that seemed overly complex.

Reply:
oh, I see :blush: this is C, not CUDA. Done.

Reply:
An easier way is to do `1.f` - this declares a float literal (`1.` is a double literal)